### PR TITLE
fix(sqlalchemy): fix correlated subqueries on projections

### PR DESCRIPTION
**Pull request by cpcloud**
While adding test coverage for ExistsSubquery and NotExistsSubquery I discovered that basic execution of correlated subqueries does not work. This PR fixes that. A couple of tests are still failing, hence the draft.

### CODE REVIEW

File: ibis/backends/base/sql/alchemy/query_builder.py [hunk 1]
```diff
@@ -227,32 +241,7 @@ def _add_select(self, table_set):
...
-        #
-        # previously this was `replace_selectable`, but that's deprecated so we
-        # inline its implementation here
-        #
-        # sqlalchemy suggests using the functionality in sa.sql.visitors, but
-        # that would effectively require reimplementing ClauseAdapter
-        replaced = sa.sql.util.ClauseAdapter(table_set).traverse(result)
-        num_froms = len(replaced.get_final_froms())
-        assert num_froms == 1, f"num_froms == {num_froms:d}"
-        return replaced
```
Comment (gforsyth) [inline ibis/backends/base/sql/alchemy/query_builder.py:255]: I am happy to see this go!

File: ibis/backends/base/sql/alchemy/registry.py [hunk 2]
```diff
@@ -81,24 +81,24 @@ def get_sqla_table(ctx, table):
...
     except KeyError:
-        return sa.column(colname)
+        # XXX: cols is a sqlalchemy column collection which contains column
+        # names that are secretly prefixed by their containing table
+        #
+        # sqlalchemy doesn't let you select by the *un*prefixed column name
+        # despite the uniqueness of `colname`
+        #
+        # however, in ibis we have already deduplicated column names so we can
+        # refer to the name by position
```
Comment (gforsyth) [inline ibis/backends/base/sql/alchemy/registry.py:99]:
```suggestion
        # cols is a sqlalchemy column collection which contains column
        # names that are secretly prefixed by their containing table
        #
        # sqlalchemy doesn't let you select by the *un*prefixed column name
        # despite the uniqueness of `colname`
        #
        # however, in ibis we have already deduplicated column names so we can
        # refer to the name by position
```

The micro-ist of nits, but to me `# XXX` reads like something that needs fixing before merge, but that is not a blocker at all.

**Review by gforsyth — APPROVED** [review]
This is a nice set of changes.

**Comment by cpcloud** [comment]
Turns out this was even more broken in the impala backend. I've got a fix that I'll push up if the test suite passes locally.

File: ibis/backends/base/sql/alchemy/query_builder.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/query_builder.py:255]: Good riddance!

File: ibis/backends/base/sql/alchemy/registry.py (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/registry.py:99]: Fixed!

**Comment by cpcloud** [comment]
Ok, so to fix this I had to completely eliminate our alias eliding, which makes the SQL slightly worse. I don't think that matters here since it is now correct and verbose whereas before it was incorrect and concise.

**Comment by cpcloud** [comment]
Actually, to make this less annoying to get in, I'm going to mark the stuff I just fixed as broken and do an immediate follow up.

**Pull request merged**